azurerm/internal/services/eventgrid/eventgrid_event_subscription_resource.go

```diff
@@ -144,6 +152,54 @@ func resourceArmEventGridEventSubscription() *schema.Resource {
 				},
 			},
 
+			"service_bus_queue_endpoint": {
```

Comment:
Given that we can only have 1 endpoint of 1 type, with 1 property, would it make sense to flatten these to: `service_bus_queue_endpoint_id`, `service_bus_topic_endpoint_id`, and `azure_function_endpoint_id` ? WDYT? unless there are known plans to add additional properties to these blocks soon not sure theres any advantage to making these named blocks with a single property

Reply:
I totally agree that this design is a bit clumsy. However, as you can see, there are several other types of endpoints all require either a resource ID, a URL, a queue name, or other endpoint-specific details. So I tried to stick to the current approach.

Reply:
I would argue that the other ones should be deprecated and simplified ie webhook_endpoint should become webhook_endpoint_url. I'm ok with `storage_queue_endpoint` being the odd one out as a block as it does have multiple properties to set for the sake of cleaning up everything else.

Reply:
Here my proposal for the resource schema:
```terraform
{
  name =  "..."
  scope = "..."

  "endpoint" {

    // Possible types:
    //   - event_hub
    //   - storage_queue
    //   - service_bus_queue
    //   - service_bus_topic
    //   - azure_function
    //   - webhook

    "type"         = "event_hub"   // Required

    // **)
    //   - event_hub
    //   - storage_queue
    //   - service_bus_queue
    //   - service_bus_topic
    //   - azure_function
    "id"           = "..."  // Optional (Only required for **)

    "webhook_url"  = "..."  // Optional (Only required for type webhook)

    "queue_name"   = "..."  // Optional (Only required for type storage_queue)
  }

}
```

Reply:
i'm not sure i'm sold on a shared endpoint block. It seems more confusing then a set of top level properties that are explity in their requirements and can easily be validated during plan (including the required properties for the block one and ensuring the ID can be parsed correctly for that type of resoruce) vs a a block where some are required as well as a new property defining the type and a id validation that changes depending on the type. Seems like a lot of added complexity for a more confusing ux?

Reply:
ah ok. now I got it, this wasn't my favorite either. I just thought you were concerned with the redundancy of the ID attributes. But it was just about nesting in separate blocks for most of the endpoint types except the storage_account_queue, as it has several attributes to be defined. sorry for my delay ... 🤦

Reply:
Hey @katbyte,

after investigating the latest EventGrid model definition inside the *azure-sdk-for-go* i'm not sure if flattening the endpoint blocks is really a good approach (beside the fact that this would be a breaking change and a lot of issue tickets would be the result).

For example the webhook endpoint supports further configuration options, which are currently not implemented within the azurerm provider resource (batch size):
https://github.com/Azure/azure-sdk-for-go/blob/master/services/preview/eventgrid/mgmt/2020-04-01-preview/eventgrid/models.go#L8168

As there were a lot of updates since the latest *azure-sdk-for-go* release i would think that it's really likely that the other endpoints could also become some more configuration options in the future.

WDYT?

Reply:
If there are additional properties not exposed for config then yea makes sense to leave it as a block! but one which are a single property i'd rather start without the block and then if/when they do add more covert it into a block (also means people will have to be aware of that change)

Theres no need for you to update the existing ones if you don't want, but for schema changes like this what we do is deprecate the old ones, conflict them with the new property then set both on read, and on create get the old one, and if the new one is populated, get and use that instead

Reply:
ah, ok, that's how it makes sense for me. I'll add the changes to have this "Deprecated / ConflictWith" stuff in it, and hopefully we'll get this PR done soon. In addition to the `storage_queue_endpoint` block, the `webhook_endpoint` block will also remain unchanged, as it makes sense to include support for batch configuration properties in the near future.